pkgs/development/python-modules/firedrake/default.nix

```
@@ -0,0 +1,180 @@
+{
+  lib,
+  stdenv,
+  buildPythonPackage,
+  fetchFromGitHub,
+  fetchpatch2,
+  python,
+  glibc,
+
+  # build-system
+  setuptools,
+  cython,
+  pybind11,
+
+  # dependencies
+  decorator,
+  cachetools,
+  mpi4py,
+  fenics-ufl,
+  firedrake-fiat,
+  h5py,
+  libsupermesh,
+  loopy,
+  petsc4py,
+  numpy,
+  packaging,
+  pkgconfig,
+  progress,
+  pyadjoint-ad,
+  pycparser,
+  pytools,
+  requests,
+  rtree,
+  scipy,
+  sympy,
+  islpy,
+  matplotlib,
+
+  # tests
+  pytest,
+  mpi-pytest,
+  mpiCheckPhaseHook,
+  writableTmpDirAsHomeHook,
+}:
+let
+  mpi4py' = mpi4py.override { inherit (petsc4py.petscPackages) mpi; };
+  h5py' = h5py.override {
+    inherit (petsc4py.petscPackages) hdf5;
+    mpi4py = mpi4py';
+  };
+  mpi-pytest' = mpi-pytest.override { mpi4py = mpi4py'; };
+in
+buildPythonPackage rec {
+  pname = "firedrake";
+  version = "2025.4.0.post0";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "firedrakeproject";
+    repo = "firedrake";
+    tag = version;
+    hash = "sha256-wQOS4v/YkIwXdQq6JMvRbmyhnzvx6wj0O6aszNa5ZMw=";
+  };
+
+  patches = [
+    (fetchpatch2 {
+      url = "https://github.com/firedrakeproject/firedrake/pull/4298/commits/194e0dc16cf6e4f29a0747424427e35d319be39f.patch?full_index=1";
+      hash = "sha256-jtCSzga8eRuAL+wjSldceG6dllRUGWuI7rpFHYbe0rE=";
+    })
+  ];
+
+  postPatch =
+    ''
+      # relax build-dependency petsc4py
+      substituteInPlace pyproject.toml --replace-fail \
+        "petsc4py==3.23.0" "petsc4py"
+
+      # These scripts are used by official source distribution only,
+      # and do not make sense in our binary distribution.
+      sed -i '/firedrake-\(check\|status\|run-split-tests\)/d' pyproject.toml
+    ''
+    + lib.optionalString stdenv.hostPlatform.isLinux ''
+      substituteInPlace firedrake/petsc.py --replace-fail \
+        'program = ["ldd"]' \
+        'program = ["${lib.getBin glibc}/bin/ldd"]'
```

```suggestion
        'program = ["${lib.getExe' glibc "ldd"}"]'
```
You can directly retrieve a specific program out of a derivation, which had some basic failsafes